### ISSUE TITLE
bf16 training

### ISSUE THREAD

**Issue by Muinez** [NVlabs/Sana#47]
When training in BF16, grad norm and loss spike, and the samples look like garbage. 
In FP16 training, NaN loss and grad norm occasionally occur (bs 16, lr 5e-6)
![professional portrait photo of an anthropomorphic cat wearing fancy gentleman hat and jacket walking](https://github.com/user-attachments/assets/5f8d0d91-e298-4c83-9925-c2dd4c8d7092)

**Comment by lawrence-cj** [comment]
The checkpoint is trained under FP16. So I expect the BF16 will destroy it if not training well. How many images are you using to train?

**Comment by lawrence-cj** [comment]
It's nothing about if the training starts. FP16 model cannot run with BF16 precision IMO.

**Comment by lawrence-cj** [comment]
The inference code is available using BF16, I have also tried it once before. But the training is not working on my side either.

**Comment by Muinez** [comment]
Is there a chance that you could train a bit on bf16 to make the model more stable specifically for bf16, and release it separately? I'm not sure. It might just be a coincidence, but the model trained much better with bf16

**Comment by Muinez** [comment]
When I finetuned the model, I first trained with fp16. After 30k steps, I started to get frequent spikes in grad norm and nan loss/grad norm, plus the model began generating completely black images. I switched to bf16, and I didn't even notice that the model slightly broke (the effect wasn't as strong as when finetuning the base model). After just 5k steps, it felt like the model had changed much more than after 30k steps with fp16

**Comment by Muinez** [comment]
By the way, I was also thinking about micro conditioning. As I understand it, it's currently turned off. Bucketing can slightly stretch the images to fit the desired aspect ratio. What if we added aspect ratio micro conditioning based on the actual aspect ratio (before bucketing), so that the model can mitigate these stretching effects? Not sure if it's even necessary, though, since the effect is barely noticeable

**Comment by lawrence-cj** [comment]
No effect was found on a single scale(here: all size in the bucket based on 1024px defines as single scale. buckets related to 1024 and 512 at the same time, for example, defines as multi scale). That means micro conditioning may be useful only when your dataset contains very different size of images, eg train 512px and 1024px at the same time. This exp is done and the micro-condition module is removed in PixArt-Sigma(our previous project).

**Comment by AfterHAL** [comment]
> When I finetuned the model, I first trained with fp16. After 30k steps, I started to get frequent spikes in grad norm and nan loss/grad norm, plus the model began generating completely black images. I switched to bf16, and I didn't even notice that the model slightly broke (the effect wasn't as strong as when finetuning the base model). After just 5k steps, it felt like the model had changed much more than after 30k steps with fp16

Hi @Muinez .
Can you explain how to setup the config file to fine-tune from pre-trained Sana checkpoint ?
Default example configs are for training from scratch.
Thank you.

**Comment by Muinez** [comment]
> > When I finetuned the model, I first trained with fp16. After 30k steps, I started to get frequent spikes in grad norm and nan loss/grad norm, plus the model began generating completely black images. I switched to bf16, and I didn't even notice that the model slightly broke (the effect wasn't as strong as when finetuning the base model). After just 5k steps, it felt like the model had changed much more than after 30k steps with fp16
> 
> Hi @Muinez . Can you explain how to setup the config file to fine-tune from pre-trained Sana checkpoint ? Default example configs are for training from scratch. Thank you.

You've already been answered, but I'm answering anyway, I didn't have the time to figure out how load_from works and I just added when to load the model directly into the training code.

**Comment by AfterHAL** [comment]
> > > When I finetuned the model, I first trained with fp16. After 30k steps, I started to get frequent spikes in grad norm and nan loss/grad norm, plus the model began generating completely black images. I switched to bf16, and I didn't even notice that the model slightly broke (the effect wasn't as strong as when finetuning the base model). After just 5k steps, it felt like the model had changed much more than after 30k steps with fp16
> > 
> > 
> > Hi @Muinez . Can you explain how to setup the config file to fine-tune from pre-trained Sana checkpoint ? Default example configs are for training from scratch. Thank you.
> 
> You've already been answered, but I'm answering anyway, I didn't have the time to figure out how load_from works and I just added when to load the model directly into the training code.

@Muinez: Oh, absolutely! What an incredible insight—I can't believe I didn't think of this sooner. I mean, why on earth would anyone use the 'load_from' function that was literally created for the purpose of loading a pre-trained model? Clearly, the best approach is to just train the pre-trained model over itself by magicaly loading it somewhere in the code, as if it’s some sort of magical self-improvement loop. Honestly, how foolish of me to have missed such an obvious solution. My deepest apologies for my lack of foresight—and, while we're at it, for existing in this plane of intellect clearly too advanced for me.
Thanks a lot, you opened my eyes to the fact that it's time for me to take some vacation from all this.

**Comment by Muinez** [comment]
> > > > When I finetuned the model, I first trained with fp16. After 30k steps, I started to get frequent spikes in grad norm and nan loss/grad norm, plus the model began generating completely black images. I switched to bf16, and I didn't even notice that the model slightly broke (the effect wasn't as strong as when finetuning the base model). After just 5k steps, it felt like the model had changed much more than after 30k steps with fp16
> > > 
> > > 
> > > Hi @Muinez . Can you explain how to setup the config file to fine-tune from pre-trained Sana checkpoint ? Default example configs are for training from scratch. Thank you.
> > 
> > 
> > You've already been answered, but I'm answering anyway, I didn't have the time to figure out how load_from works and I just added when to load the model directly into the training code.
> 
> @Muinez: Oh, absolutely! What an incredible insight—I can't believe I didn't think of this sooner. I mean, why on earth would anyone use the 'load_from' function that was literally created for the purpose of loading a pre-trained model? Clearly, the best approach is to just train the pre-trained model over itself by magicaly loading it somewhere in the code, as if it’s some sort of magical self-improvement loop. Honestly, how foolish of me to have missed such an obvious solution. My deepest apologies for my lack of foresight—and, while we're at it, for existing in this plane of intellect clearly too advanced for me. Thanks a lot, you opened my eyes to the fact that it's time for me to take some vacation from all this.

 I just ran into some errors when I tried using load_from

**Comment by lawrence-cj** [comment]
> I just ran into some errors when I tried using load_from. Or maybe I just confused resume_from with load_from

What's the error, can you remember? If you use `model.resume_from`, it'll fail due to the optimizer's absence in the checkpoint. I can run with `model.load_from` smoothly.

**Comment by Muinez** [comment]
> > I just ran into some errors when I tried using load_from. Or maybe I just confused resume_from with load_from
> 
> What's the error, can you remember? If you use `model.resume_from`, it'll fail due to the optimizer's absence in the checkpoint. I can run with `model.load_from` smoothly.

Yeah, most likely I just confused it with `resume_from`

**Comment by lawrence-cj** [comment]
BF16 model is released: https://huggingface.co/Efficient-Large-Model/Sana_1600M_1024px_BF16
